File: java/selenium/src/test/java/dev/selenium/request/MultipleHandlersTest.java
package dev.selenium.request;

import dev.selenium.TestBase;
import java.net.URI;
import java.util.function.Predicate;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.openqa.selenium.firefox.FirefoxDriver;

public class MultipleHandlersTest extends TestBase {

  @Test
  public void modifyMultiple() {
    Predicate<URI> filter = uri -> uri.getHost().equals("httpbin.org");

    ((FirefoxDriver) driver)
        .network()
        .addRequestHandler(filter, req -> req.removeHeader("upgrade-insecure-requests"));

    ((FirefoxDriver) driver)
        .network()
        .addRequestHandler(filter, req -> req.addHeader("X-Test", "true"));

    //    UrlPattern filter = UrlPatternBuilder.setHost("selenium.dev").build();
    //    driver
    //        .network()
    //        .addRequestHandler(filter, route -> route.next(route.request.addHeader("X-Test",
    // "true")));
    //
    //    driver
    //        .network()
    //        .addRequestHandler(
    //            filter, req ->
    // route.next(route.request.removeHeader("upgrade-insecure-requests")));

    driver.get("https://httpbin.org/headers");

    String bodyText =
        driver.findElement(org.openqa.selenium.By.tagName("body")).getText().replaceAll("\\s+", "");
    Assertions.assertFalse(bodyText.contains("Upgrade-Insecure-Requests"));
    Assertions.assertTrue(
        bodyText.contains("\"X-Test\":\"true\""), "Does not yet support multiple handlers");
  }
}

File: dotnet/Selenium/Request/MultipleHandlersTest.cs
using Microsoft.VisualStudio.TestTools.UnitTesting;
using OpenQA.Selenium;
using OpenQA.Selenium.BiDi.Network;

namespace Selenium.Requests;

[TestClass]
public class MultipleHandlersTest : TestBase
{
    [TestMethod]
    public async Task ModifyMultiple()
    {
        await using var intercept1 = await Bidi.Network.InterceptRequestAsync(async req =>
        {
            var headers = new List<Header>(req.Request.Headers) { new("X-Test", "true") };

            await req.ContinueAsync(new ContinueRequestOptions { Headers = headers });
        });

        await using var intercept2 = await Bidi.Network.InterceptRequestAsync(async req =>
        {
            var headers = new List<Header?>(req.Request.Headers);
            var remove = headers.Find(h => h?.Name == "Upgrade-Insecure-Requests");
            headers.Remove(remove);

            await req.ContinueAsync(new ContinueRequestOptions { Headers = headers! });
        });

        // var options = new RequestInterceptOptions {
        //               UrlPatterns = new List<UrlPattern> { new UrlPattern("https://httbin.org/") },
        //           };
        // await using var registration1 = await driver.Network.AddRequestHandlerAsync(
        //     options,
        //     req =>
        //     {
        //          req.RemoveHeader("Upgrade-Insecure-Requests");
        //      }
        // );

        // await using var registration2 = await driver.Network.AddRequestHandlerAsync(
        //     options,
        //     req =>
        //     {
        //          req.AddHeader("X-Test", "true");
        //      }
        // );

        await Driver.Navigate().GoToUrlAsync("https://httpbin.org/headers");

        Assert.AreEqual(0, Driver.FindElements(By.Id("/headers/Upgrade-Insecure-Requests")).Count,
            "Does not support multiple intercepts");
        Assert.AreEqual("X-Test \"true\"", Driver.FindElement(By.Id("/headers/X-Test")).Text,
            "Does not support multiple intercepts");
    }
}